Lib/imaplib.py

```diff
@@ -315,14 +331,58 @@ def open(self, host='', port=IMAP4_PORT, timeout=None):
 
     def read(self, size):
         """Read 'size' bytes from remote."""
-        return self.file.read(size)
+        # Read from an unbuffered input, so our select() calls will not be
+        # defeated by a hidden library buffer.  Use our own buffer instead,
+        # which can be examined before calling select().
+        if isinstance(self, IMAP4_stream):
+            read = self.readfile.read
+        else:
+            read = self.sock.recv
```

Comment:
Out of curiosity, can't we have a private `_read()` that would be defined to be `self.readfile.read` on IMAP4_stream classes and `self.sock.recv` otherwise? This would allow us to avoid this kind of type checks.